lib/babe/build.go

```diff
@@ -189,52 +189,78 @@ func (b *BlockBuilder) buildBlockSeal(header *types.Header) (*types.SealDigest,
 func (b *BlockBuilder) buildBlockExtrinsics(slot Slot, rt runtime.Instance) []*transaction.ValidTransaction {
 	var included []*transaction.ValidTransaction
 
-	for !hasSlotEnded(slot) {
-		txn := b.transactionState.Pop()
-		// Transaction queue is empty.
-		if txn == nil {
-			continue
+	slotEnd := slot.start.Add(slot.duration * 2 / 3) // reserve last 1/3 of slot for block finalisation
+	timeout := time.Until(slotEnd)
+	slotTimer := time.NewTimer(timeout)
+
+	for {
+		select {
+		case <-slotTimer.C:
+			return included
+		default:
 		}
 
-		extrinsic := txn.Extrinsic
-		logger.Tracef("build block, applying extrinsic %s", extrinsic)
+		popChan, cancel := b.transactionState.PopChannel()
+		defer func() {
```

Comment:
we probably should not rely on `defer` since this would stack N defers for N iterations of that for loop.

Reply:
Addressed by refactoring how popChan is created and handled.